### PR TITLE
Move classes used for generating Groovy interceptors from :core to :internal-instrumentation-api

### PR DESCRIPTION
**Pull request by asodja**
So we can upgrade classes in core-api.

Fixes https://github.com/gradle/gradle/issues/29239

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

**Comment by asodja** [comment]
@bot-gradle test this

**Comment by asodja** [comment]
@bot-gradle clean

File: subprojects/core/src/test/groovy/org/gradle/internal/classpath/intercept/InterceptScopeTest.groovy
Comment (mlopatkin) [inline subprojects/core/src/test/groovy/org/gradle/internal/classpath/intercept/InterceptScopeTest.groovy:1]: 💅  I think this test should be moved too, alongside the class its tests.

File: subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptorResolver.java
Comment (mlopatkin) [inline subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptorResolver.java:1]: ❓ Why is the reset of the package still in `:core`?

File: platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/groovy/InterceptGroovyCallsGenerator.java [hunk 1]
```diff
@@ -275,12 +281,12 @@ private static Type propertyReceiverType(CallInterceptionRequest request) {
...
-    private static final ClassName CALL_INTERCEPTOR_CLASS = ClassName.bestGuess("org.gradle.internal.classpath.intercept.AbstractCallInterceptor");
-    private static final ClassName SIGNATURE_AWARE_CALL_INTERCEPTOR_CLASS = ClassName.bestGuess("org.gradle.internal.classpath.intercept.SignatureAwareCallInterceptor");
-    private static final ClassName SIGNATURE_AWARE_CALL_INTERCEPTOR_SIGNATURE_MATCH =
-        ClassName.bestGuess("org.gradle.internal.classpath.intercept.SignatureAwareCallInterceptor.SignatureMatch");
-    private static final ClassName PROPERTY_AWARE_CALL_INTERCEPTOR_CLASS = ClassName.bestGuess("org.gradle.internal.classpath.intercept.PropertyAwareCallInterceptor");
-    private static final ClassName INTERCEPTED_SCOPE_CLASS = ClassName.bestGuess("org.gradle.internal.classpath.intercept.InterceptScope");
-    private static final ClassName INVOCATION_CLASS = ClassName.bestGuess("org.gradle.internal.classpath.intercept.Invocation");
+    static final ClassName FILTERABLE_CALL_INTERCEPTOR = ClassName.get(FilterableCallInterceptor.class);
```
Comment (mlopatkin) [inline platforms/core-runtime/internal-instrumentation-processor/src/main/java/org/gradle/internal/instrumentation/processor/codegen/groovy/InterceptGroovyCallsGenerator.java:284]: 👍

**Review by mlopatkin — CHANGES_REQUESTED** [review]
Looks good! 

For the remaining interceptor implementation classes, I think they should be moved too, the current split now feels a bit arbitrary.

File: subprojects/core/src/test/groovy/org/gradle/internal/classpath/intercept/InterceptScopeTest.groovy
Comment (asodja) [reply subprojects/core/src/test/groovy/org/gradle/internal/classpath/intercept/InterceptScopeTest.groovy:1]: Good catch

File: subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptorResolver.java
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptorResolver.java:1]: I moved only as little as is needed. I can try to move other classes too if they don't have any dependencies

**Comment by asodja** [comment]
@bot-gradle test this

File: subprojects/core/src/test/groovy/org/gradle/internal/classpath/intercept/InterceptScopeTest.groovy
Comment (asodja) [reply subprojects/core/src/test/groovy/org/gradle/internal/classpath/intercept/InterceptScopeTest.groovy:1]: Fixed

File: subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptorResolver.java
Comment (asodja) [reply subprojects/core/src/main/java/org/gradle/internal/classpath/intercept/CallInterceptorResolver.java:1]: The rest of the package is not actually used in generated classes but only at runtime. I left only classes that are responsible to discovering interceptors in `:core` for now.

I guess we could at one point add `instrumentation-runtime` project (or just move them to `instrumentation-api`). But it looks like it needs some more untangling. So I left them there for now.

**Comment by asodja** [comment]
> For the remaining interceptor implementation classes, I think they should be moved too, the current split now feels a bit arbitrary.

Moved `ClassBoundCallInterceptor`, `CompositeCallInterceptor` and `FilterableCallInterceptor` as well. Others are concrete implementation which in `Instrumented.java` which I think doesn't belong in `instrumentation-api` project.

**Comment by asodja** [comment]
@bot-gradle clean

**Review by mlopatkin — APPROVED** [review]

**Pull request merged**